[ProfileData] Add getValueArrayForSite

kazutakahirata opened the pull request:
Without this patch, a typical traversal over the value data looks
like:

  uint32_t NV = Func.getNumValueDataForSite(VK, S);
  std::unique_ptr<InstrProfValueData[]> VD = Func.getValueForSite(VK, S);
  for (uint32_t V = 0; V < NV; V++)
    Do something with VD[V].Value and/or VD[V].Count;

This patch adds getValueArrayForSite, which returns
ArrayRef<InstrProfValueData>, so we can do:

  for (const auto &V : Func.getValueArrayForSite(VK, S))
    Do something with V.Value and/or V.Count;

I'm planning to migrate the existing uses of getValueForSite to
getValueArrayForSite in follow-up patches and remove getValueForSite
and getNumValueDataForSite.

kazutakahirata committed:
[ProfileData] Teach getValueForSite to return ArrayRef<InstrProfValueData>

Without this patch, a typical traversal over the value data looks
like:

  uint32_t NV = Func.getNumValueDataForSite(VK, S);
  std::unique_ptr<InstrProfValueData[]> VD = Func.getValueForSite(VK, S);
  for (uint32_t V = 0; V < NV; V++)
    Do something with VD[V].Value and/or VD[V].Count;

With this patch, getValueForSite returns ArrayRef<InstrProfValueData>,
so we can do:

  for (const auto &V : Func.getValueForSite(VK, S))
    Do something with V.Value and/or V.Count;

To accommodate the migration, this patch renames the existing
getValueForSite and its uses to getValueForSiteLegacy.  This patch
silently switches to the new signature for unit tests that don't
require any change like:

  auto VD0 = R->getValueForSite(IPVK_VTableTarget, 0);
  EXPECT_EQ(VD0[0].Value, getCalleeAddress(vtable2));
  EXPECT_EQ(VD0[0].Count, 7U);

where we receive the return value with auto and reference the array
contents.

Everything else uses getValueForSiteLegacy for now.  I'm planning to
migrate to the new signature and remove getValueForSiteLegacy and
getNumValueDataForSite in follow-up patches.

kazutakahirata committed:
Fix formatting.

minglotus-6 commented on llvm/include/llvm/ProfileData/InstrProf.h:
```diff
@@ -869,14 +869,18 @@ struct InstrProfRecord {
+  inline ArrayRef<InstrProfValueData> getValueForSite(uint32_t ValueKind,
+                                                      uint32_t Site) const;
+
   /// Return the array of profiled values at \p Site.
```
The new interface returns an ArrayRef and the deprecated interface allows caller to take ownership of the array.

I wonder if there are existing call sites where explicit ownership is preferred over a reference (e.g., easy to reason about lifecycle, etc).

minglotus-6 replied:
If yes I think we may need both.

kazutakahirata replied:
I wouldn't worry about the lifetime of profile data.  The profile data stays available at least during the entire duration of a pass.

mingmingl-llvm reviewed (commented):
thanks!

minglotus-6 commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -2696,7 +2696,8 @@ static void traverseAllValueSites(const InstrProfRecord &Func, uint32_t VK,
   for (size_t I = 0; I < NS; ++I) {
     uint32_t NV = Func.getNumValueDataForSite(VK, I);
-    std::unique_ptr<InstrProfValueData[]> VD = Func.getValueForSite(VK, I);
+    std::unique_ptr<InstrProfValueData[]> VD =
```
ref: 'Local variable type deduction' in https://google.github.io/styleguide/cppguide.html

kazutakahirata replied:
Thanks!  The latest iteration does not touch this part of the code.  I'll keep `auto VD` in mind when I migrate the code in question to `getValueArrayForSite` in follow-up patches.

david-xl reviewed (commented):
I suggest leaving the old interface name unchanged and make the new interface getValueArrayForSite.

kazutakahirata committed:
Use getValueArrayForSite instead of getValueForSite.

kazutakahirata commented on llvm/lib/ProfileData/InstrProfWriter.cpp:
```diff
@@ -1090,7 +1091,8 @@ void InstrProfWriter::writeRecordInText(StringRef Name, uint64_t Hash,
       uint32_t ND = Func.getNumValueDataForSite(VK, S);
       OS << ND << "\n";
-      std::unique_ptr<InstrProfValueData[]> VD = Func.getValueForSite(VK, S);
+      std::unique_ptr<InstrProfValueData[]> VD =
```
Likewise, the latest iteration does not touch this part of the code.  I'll keep `auto VD` in mind when I migrate the code in question to `getValueArrayForSite` in follow-up patches.

kazutakahirata commented:
> I suggest leaving the old interface name unchanged and make the new interface getValueArrayForSite.

Fixed in the latest iteration and updated the title and commit comment accordingly.

PTAL again.  Thanks!

kazutakahirata committed:
Fix formatting.
